src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -674,3 +682,430 @@ def _standard_metadata(self, name: str) -> V1ObjectMeta:
+            except asyncio.TimeoutError:
+                self._logger.error(
+                    f"Fileserver for {username} did not appear."
+                )
+                return False
+
+    async def check_namespace(self, namespace: str) -> bool:
+        """Check to see if namespace is present; return True if it is,
+        False if it is not."""
+        try:
+            await self.api.read_namespace(namespace)
+            return True
+        except ApiException as e:
+            if e.status != 404:
+                msg = f"Cannot read namespace {namespace}"
+                raise KubernetesError.from_exception(
+                    msg, e, namespace=namespace
+                ) from e
+            return False
+
+    async def _create_fileserver_job(
+        self, username: str, namespace: str, job: V1Job
+    ) -> None:
+        obj_name = f"{username}-fs"
+        self._logger.debug("Creating job", name=obj_name, namespace=namespace)
+        try:
+            await self.batch_api.create_namespaced_job(namespace, job)
+        except ApiException as e:
+            if e.status == 409:
+                # It already exists.  Delete and recreate it
+                self._logger.warning(
+                    "Job exists.  Deleting and recreating.",
+                    name=obj_name,
+                    namespace=namespace,
+                )
+                await self._delete_fileserver_job(username, namespace)
+                await self._wait_for_fileserver_object_deletion(
+                    obj_name=obj_name, namespace=namespace, kind="job"
+                )
+                await self.batch_api.create_namespaced_job(namespace, job)
+                return
+            raise KubernetesError.from_exception(
+                "Error creating job",
+                e,
+                namespace=namespace,
+                name=obj_name,
+            ) from e
+
+    async def _delete_fileserver_job(
+        self, username: str, namespace: str
+    ) -> None:
+        obj_name = f"{username}-fs"
+        self._logger.debug("Deleting job", name=obj_name, namespace=namespace)
+        try:
+            await self.batch_api.delete_namespaced_job(
+                obj_name, namespace, propagation_policy="Foreground"
+            )
+        except ApiException as e:
+            if e.status == 404:
+                return
+            raise KubernetesError.from_exception(
+                "Error deleting job",
+                e,
+                namespace=namespace,
+                name=obj_name,
+            ) from e
+
+    async def _create_fileserver_service(
+        self, username: str, namespace: str, spec: V1Service
+    ) -> None:
+        obj_name = f"{username}-fs"
+        try:
+            await self.api.create_namespaced_service(namespace, spec)
+        except ApiException as e:
+            if e.status == 409:
+                # It already exists.  Delete and recreate it
```

Comment:
Likewise here, not sure why deleting and recreating on conflict is the right thing to do.

Reply:
comments added.